duplex_http_call.go

```diff
@@ -113,18 +123,51 @@ func (d *duplexHTTPCall) Send(payload messsagePayload) (int64, error) {
 		// Signal that the stream is closed with the more-typical io.EOF instead of
 		// io.ErrClosedPipe. This makes it easier for protocol-specific wrappers to
 		// match grpc-go's behavior.
-		return bytesWritten, io.EOF
+		err = io.EOF
 	}
 	return bytesWritten, err
 }
 
+func (d *duplexHTTPCall) sendUnary(payload messagePayload) (int64, error) {
+	// Unary messages are sent as a single HTTP request. We don't need to use a
+	// pipe for the request body and we don't need to send headers separately.
+	if !d.requestSent.CompareAndSwap(false, true) {
+		return 0, fmt.Errorf("request already sent")
+	}
+	payloadLength := int64(payload.Len())
+	if payloadLength > 0 {
+		// Build the request body from the payload.
+		payloadBody := newPayloadCloser(payload)
+		d.request.Body = payloadBody
+		d.request.ContentLength = payloadLength
+		d.request.GetBody = func() (io.ReadCloser, error) {
+			if !payloadBody.Rewind() {
+				return nil, fmt.Errorf("payload cannot be retried")
+			}
+			return payloadBody, nil
+		}
+		// Wait on the payloadBody to be completly read or closed before
+		// returning from Send. This ensures that the payload can be reused
+		// after Send returns. See [http.RoundTripper] for more details.
+		defer payloadBody.Wait()
+	}
+	d.makeRequest() // synchronous request
+	if err := d.ctx.Err(); err != nil {
+		return 0, wrapIfContextError(err)
+	}
```

Comment:
This seems like it has the potential for returning a cancelled or deadline exceeded error even if the call succeeded -- like if a concurrent cancelation happens between the actual call inside of `makeRequest` and this check. Can we instead only do this if there was an error making the request (which should be do-able since it's synchronous)?

Reply:
We can surface errors here but it will break tests as current behaviour expects requests errors to be surfaced on reading the response. I'd like to leave as an optimization for afterwards.

Reply:
That's not at all what I meant -- I did not suggest that we start returning new errors here. What I meant was that we _not_ return a cancelled error here unless there was _actually_ an error in `makeRequest`. If the HTTP call was successful as far as we know, let any cancellations/timeouts be discovered in calls to `Receive` the response.

In particular, this is the sequence I'm concerned about:
1. HTTP call is made.
2. HTTP call is successful, with legit response.
3. Context is cancelled.
4. We notice here that context is cancelled and prevent access to successfully completed response.

In the above case, if the cancellation did not actually prevent us from receiving a successful response, we should allow the caller to access that response instead of returning a context error.

Reply:
Ah right. For the `2.` -> `3.` case this will only happen in a very small time window. I don't see the advantage in making the error check more complex for a use case that occurs on a race between completion and context close. I'd think knowing about a context cancel earlier is also beneficial as any call to read the request will also check the context first blocking access to the request.

Reply:
Fixed in 12c36b0